File: frontend/src/lib/api/tryon.ts
/**
 * Try-On API functions for FitView AI.
 * Phase 3: Core Virtual Try-On Engine.
 */

const API_BASE = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000/api/v1";

export interface TryOnRequest {
  model_id: string;
  product_id: string;
}

export interface TryOnSession {
  id: string;
  _id: string;
  user_id: string;
  model_id: string;
  product_id: string;
  result_url: string;
  model_name: string;
  product_name: string;
  model_image_url: string;
  product_image_url: string;
  status: "pending" | "processing" | "completed" | "failed";
  processing_time_ms: number;
  is_favorite: boolean;
  created_at: string;
  expires_at: string | null;
}

export interface TryOnHistoryResponse {
  sessions: TryOnSession[];
  total: number;
  page: number;
  limit: number;
}

function getAuthHeaders(): HeadersInit {
  const token = typeof window !== "undefined" ? localStorage.getItem("token") : null;
  const headers: Record<string, string> = {
    "Content-Type": "application/json",
  };
  if (token) {
    headers["Authorization"] = `Bearer ${token}`;
  }
  return headers;
}

export async function generateTryOn(data: TryOnRequest): Promise<TryOnSession> {
  const res = await fetch(`${API_BASE}/tryon`, {
    method: "POST",
    headers: getAuthHeaders(),
    body: JSON.stringify(data),
  });
  if (!res.ok) {
    const error = await res.json().catch(() => ({ detail: "Try-on generation failed" }));
    throw new Error(error.detail || "Try-on generation failed");
  }
  return res.json();
}

export async function getTryOnHistory(
  page: number = 1,
  limit: number = 20
): Promise<TryOnHistoryResponse> {
  const params = new URLSearchParams();
  params.set("page", String(page));
  params.set("limit", String(limit));

  const res = await fetch(`${API_BASE}/tryon/history?${params.toString()}`, {
    headers: getAuthHeaders(),
  });
  if (!res.ok) {
    const error = await res.json().catch(() => ({ detail: "Failed to fetch history" }));
    throw new Error(error.detail || "Failed to fetch history");
  }
  return res.json();
}

export async function getTryOnSession(sessionId: string): Promise<TryOnSession> {
  const res = await fetch(`${API_BASE}/tryon/${sessionId}`, {
    headers: getAuthHeaders(),
  });
  if (!res.ok) {
    const error = await res.json().catch(() => ({ detail: "Failed to fetch session" }));
    throw new Error(error.detail || "Failed to fetch session");
  }
  return res.json();
}

export async function toggleTryOnFavorite(
  sessionId: string,
  isFavorite: boolean
): Promise<TryOnSession> {
  const res = await fetch(`${API_BASE}/tryon/${sessionId}/favorite`, {
    method: "PATCH",
    headers: getAuthHeaders(),
    body: JSON.stringify({ is_favorite: isFavorite }),
  });
  if (!res.ok) {
    const error = await res.json().catch(() => ({ detail: "Failed to update favorite" }));
    throw new Error(error.detail || "Failed to update favorite");
  }
  return res.json();
}

File: backend/app/models/tryon.py
"""
Try-On Pydantic schemas for FitView AI.
Phase 3: Core Virtual Try-On Engine.
"""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class TryOnRequest(BaseModel):
    """Schema for requesting a virtual try-on generation."""
    model_config = {"protected_namespaces": ()}

    model_id: str = Field(..., description="Fashion model ID")
    product_id: str = Field(..., description="Product/garment ID")


class TryOnResponse(BaseModel):
    """Schema for try-on generation response."""
    model_config = {
        "populate_by_name": True,
        "protected_namespaces": (),
        "json_encoders": {datetime: lambda v: v.isoformat()},
    }

    id: str = Field(..., alias="_id")
    user_id: str
    model_id: str
    product_id: str
    result_url: str = Field(..., description="URL of the generated try-on image")
    model_name: str = ""
    product_name: str = ""
    model_image_url: str = ""
    product_image_url: str = ""
    status: str = Field(default="completed", description="pending | processing | completed | failed")
    processing_time_ms: int = Field(default=0, description="Time taken to generate in milliseconds")
    is_favorite: bool = False
    ai_provider: str = Field(default="fallback", description="AI provider used: gemini or fallback")
    created_at: datetime
    expires_at: Optional[datetime] = None


class TryOnHistoryResponse(BaseModel):
    """Schema for paginated try-on history response."""
    sessions: list[TryOnResponse]
    total: int
    page: int
    limit: int


class TryOnFavoriteRequest(BaseModel):
    """Schema for toggling favorite on a try-on session."""
    is_favorite: bool


class BatchTryOnRequest(BaseModel):
    """Schema for requesting batch virtual try-on generation (multiple garments)."""
    model_config = {"protected_namespaces": ()}

    model_id: str = Field(..., description="Fashion model ID")
    product_ids: list[str] = Field(
        ...,
        min_length=1,
        max_length=5,
        description="List of product/garment IDs (1-5)",
    )


class BatchTryOnResponse(BaseModel):
    """Schema for batch try-on generation response."""
    batch_id: str
    individual_results: list[TryOnResponse]
    combined_result: Optional[TryOnResponse] = None
    total_processing_time_ms: int
    product_count: int
